src/pip/_internal/index/package_finder.py

```diff
@@ -452,24 +452,22 @@ def get_applicable_candidates(
         # Using None infers from the specifier instead.
         allow_prereleases = self._allow_all_prereleases or None
         specifier = self._specifier
-        versions = {
-            str(v)
-            for v in specifier.filter(
+        candidates_and_versions = [(c, str(c.version)) for c in candidates]
+        versions = set(
+            specifier.filter(
                 # We turn the version object into a str here because otherwise
                 # when we're debundled but setuptools isn't, Python will see
                 # packaging.version.Version and
                 # pkg_resources._vendor.packaging.version.Version as different
                 # types. This way we'll use a str as a common data interchange
                 # format. If we stop using the pkg_resources provided specifier
                 # and start using our own, we can drop the cast to str().
```

Comment:
This comment seems a bit out of place now. Perhaps it should be moved up?

Reply:
Moved